src/System.Text.Json/src/System/Text/Json/JsonParseObject.cs

```diff
@@ -0,0 +1,350 @@
+                        {
+                            length = BitConverter.ToInt32(_buffer, i + 4);
+                            newEnd = newEnd + RowSize * length;
+                        }
+                        return new JsonParseObject(_buffer, newStart, newEnd);
+                    }
+
+                    if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+                    {
+                        length = BitConverter.ToInt32(_buffer, i + 4);
+                        i += length * RowSize;
+                    }
+
+                    counter++;
+                }
+
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        public static explicit operator string (JsonParseObject json)
+        {
+            return GetUtf8String(json).ToString();
+        }
+
+        public static explicit operator Utf8String(JsonParseObject json)
+        {
+            return GetUtf8String(json);
+        }
+
+        private static Utf8String GetUtf8String(JsonParseObject json)
+        {
+            var typeCode = json._buffer[json._start + 8];
+
+            if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+            {
+                throw new InvalidCastException();
+            }
+
+            int location = BitConverter.ToInt32(json._buffer, json._start);
+            int length = BitConverter.ToInt32(json._buffer, json._start + 4);
+            return new Utf8String(json._buffer, location, length);
+        }
+
+        public static explicit operator bool (JsonParseObject json)
+        {
+            var typeCode = json._buffer[json._start + 8];
+
+            if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+            {
+                throw new InvalidCastException();
+            }
+
+            int length = BitConverter.ToInt32(json._buffer, json._start + 4);
+            if (length < 4 || length > 5)
+            {
+                throw new InvalidCastException();
+            }
+
+            int location = BitConverter.ToInt32(json._buffer, json._start);
+            bool isTrue = json._buffer[location] == 't' && json._buffer[location + 1] == 'r' && json._buffer[location + 2] == 'u' && json._buffer[location + 3] == 'e';
+            bool isFalse = json._buffer[location] == 'f' && json._buffer[location + 1] == 'a' && json._buffer[location + 2] == 'l' && json._buffer[location + 3] == 's' && json._buffer[location + 4] == 'e';
+
+            if (isTrue)
+            {
+                return true;
+            }
+            else if (isFalse)
+            {
+                return false;
+            }
+            else
+            {
+                throw new InvalidCastException();
+            }
+        }
+
+        public static explicit operator int (JsonParseObject json)
```

Comment:
This should really be in System.Text.Formatting package in the parser namespace (see InvariantParser class). BTW, I am ok with you punting these large changes for later, as long as you file issues for them.